src/main/java/graphql/schema/SchemaUtil.java

```diff
@@ -5,67 +5,48 @@
 import graphql.AssertException;
 import graphql.Internal;
 import graphql.introspection.Introspection;
+import graphql.util.TraversalControl;
+import static graphql.util.TraversalControl.QUIT;
+import graphql.util.TraverserContext;
 
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.HashMap;
-import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 import static java.lang.String.format;
 
 @Internal
 public class SchemaUtil {
 
-    public boolean isLeafType(GraphQLType type) {
-        GraphQLUnmodifiedType unmodifiedType = getUnmodifiedType(type);
-        return
-                unmodifiedType instanceof GraphQLScalarType
-                        || unmodifiedType instanceof GraphQLEnumType;
+    public Boolean isLeafType(GraphQLType graphQLType) {
+        return (Boolean) SHALLOW_TRAVERSER.depthFirst(new LeafVisitor(), graphQLType).getResult();
     }
```

Comment:
I find this one a real stretch in terms of saying this is "better"...There are 2 conditions -scalar and enum and there is some unwrapping to be done.

If I look at this code is an indirection that is low value.

That said the others have more value

Reply:
Unwrapping is using recursion, which is based on a fixed stack. 

Large and deep graphs have a chance to overflow stack. VERY large graphql type system will overflow it for sure. 

Recursion is also application-dependent at runtime. If this lib will be called after web server, IoC framework, other application layers .... this all eats up stack leaving less and less room. Not to mention hypothetical cycles in graphql type system (a.k.a schema)...

Another hypothetical  case is there application developer would run graphql in a thread with custom (and small) execution stack. 

Traverser is free from all above issues. It relies on a true stack, which would grow in size if needed and tracks visited types as it goes.

Reply:
In theory - perhaps

but in practice, given a graphql type it can only be 1 of 6 different types (enum, scalar, object, input, interface and union) or the wrappers like (list, null, type reference)

So in order to blow the stack given a single graphql type, then you would need megabytes of wrapping

eg `[![![![![![![![![![![![![![SomeTypeHere]]]]]]]]]]]`

Given a simple graphql type there is no infinite tree.

Now if you include `graphql.schema.GraphQLFieldDefinition` (which does not derive from GraphqlType) then yes you can have a large tree.

My point is that not ALL uses case of recursion are problematic.  And in this case where we have a single GraphqlType, not at all.

Reply:
> My point is that not ALL uses case of recursion are problematic.
Agree on 100%. Do you think that traverser gives too much of overhead?  

This case might be trivial, but it is still type traversal. Traverser looks good enough to cover this case too. Would you prefer to have custom algorithm here instead?